### PR TITLE
Attention Temperature Curriculum: Broad→Sharp Slice Routing

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The `temperature` parameter in `Physics_Attention_Irregular_Mesh` is initialized at 0.5 and left to learn freely. It controls how broadly/sharply the slice attention distributes point features:
- **High temperature** → broad softmax, many slices get signal (exploration)
- **Low temperature** → sharp softmax, each point commits to few slices (specialization)

The model currently finds its own temperature from random init, which may converge to a local optimum where tandem points — which need broader context about the paired foil — are over-committed too early in training.

**Proposed:** Externally schedule temperature from high (2.0) → low (0.3) over the first 80 epochs, then release to free learning. This mirrors "warm exploration then exploit" from simulated annealing. Early broad routing helps NACA6416 geometry share signal with seen geometries.

This is orthogonal to GSB (which changes *where* routing happens) — this changes *how broadly* routing happens during training. **Zero new parameters.**

## Instructions

**Step 1 — Add CLI flags:**
```python
parser.add_argument('--temp_curriculum', action='store_true',
                    help='Enable attention temperature scheduling')
parser.add_argument('--temp_init', type=float, default=2.0,
                    help='Starting temperature (high = broad)')
parser.add_argument('--temp_floor', type=float, default=0.3,
                    help='Ending temperature before release')
parser.add_argument('--temp_anneal_epochs', type=int, default=80,
                    help='Epochs over which to anneal')
```

**Step 2 — Add temperature scheduling in the training loop.** At the start of each epoch (after the epoch counter increments), inject the scheduled temperature into all attention modules:

```python
if args.temp_curriculum:
    if epoch < args.temp_anneal_epochs:
        frac = epoch / args.temp_anneal_epochs
        scheduled_temp = args.temp_init * (1 - frac) + args.temp_floor * frac
        for module in model.modules():
            if hasattr(module, 'temperature') and isinstance(module.temperature, nn.Parameter):
                module.temperature.data.fill_(scheduled_temp)
                module.temperature.requires_grad_(False)  # freeze during schedule
    elif epoch == args.temp_anneal_epochs:
        # Release: unfreeze temperature parameters
        for module in model.modules():
            if hasattr(module, 'temperature') and isinstance(module.temperature, nn.Parameter):
                module.temperature.requires_grad_(True)
    # After release, temperature learns freely (no intervention)
```

**Important:** The EMA model's temperature should NOT be externally set — let it accumulate naturally from the training model's updates. Only modify the training model's `temperature` parameter.

**Step 3 — Run 2 configs, 2 seeds each** using `--wandb_group fern-temp-curriculum`:

**Config A — standard anneal (init=2.0, floor=0.3, anneal_epochs=80):**
```bash
cd cfd_tandemfoil && python train.py --agent fern \
  --wandb_name "fern/temp-curr-2.0-0.3-80" --wandb_group fern-temp-curriculum \
  --temp_curriculum --temp_init 2.0 --temp_floor 0.3 --temp_anneal_epochs 80 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

**Config B — shorter anneal (init=1.5, floor=0.5, anneal_epochs=60):**
```bash
cd cfd_tandemfoil && python train.py --agent fern \
  --wandb_name "fern/temp-curr-1.5-0.5-60" --wandb_group fern-temp-curriculum \
  --temp_curriculum --temp_init 1.5 --temp_floor 0.5 --temp_anneal_epochs 60 \
  ... (same baseline flags) \
  --seed 42
```
(repeat with `--seed 73`)

Run all 4 in parallel. Report p_in, p_oodc, p_tan, p_re per config (2-seed avg).

**Diagnostic:** Log the temperature value every 10 epochs to W&B (both during schedule and after release) so we can see where it converges.

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Summary

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| A (2.0→0.3, 80ep) | 42 | 12.9 | 8.1 | 29.7 | 6.3 | 44.8GB | nvyisrfa |
| A (2.0→0.3, 80ep) | 73 | 13.4 | 8.2 | 29.8 | 6.5 | 46.2GB | zjx1jmy9 |
| **Config A avg** | — | **13.15** | **8.15** | **29.75** | **6.40** | — | — |
| B (1.5→0.5, 60ep) | 42 | 13.1 | 8.0 | 29.3 | 6.5 | 44.8GB | 9uii4cn2 |
| B (1.5→0.5, 60ep) | 73 | 13.6 | 8.2 | 29.4 | 6.6 | 46.2GB | k4k8ko01 |
| **Config B avg** | — | **13.35** | **8.10** | **29.35** | **6.55** | — | — |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | — | d7l91p0x, j9btfx09 |

**Negative result.** Both configs regressed on p_tan and p_oodc vs. baseline.

| Metric | Config A change | Config B change |
|--------|-----------------|-----------------|
| p_in   | +0.8%           | +2.3%           |
| p_oodc | **+5.8%** ↑ worse | **+5.2%** ↑ worse |
| p_tan  | **+4.0%** ↑ worse | **+2.6%** ↑ worse |
| p_re   | -2.3% ✓ slight improvement | 0.0% neutral |

### Commands used

```bash
# Config A, Seed 42
cd cfd_tandemfoil && python train.py --agent fern   --wandb_name "fern/temp-curr-2.0-0.3-80-s42" --wandb_group fern-temp-curriculum   --temp_curriculum --temp_init 2.0 --temp_floor 0.3 --temp_anneal_epochs 80   --seed 42   --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4   --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96   --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5   --cosine_T_max 160 --pressure_first --pressure_deep   --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3   --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05   --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Config B uses --temp_init 1.5 --temp_floor 0.5 --temp_anneal_epochs 60
```

### Training details

- Epochs: 150-151 for both configs — wall-clock limit at 180 min
- VRAM: ~44.8–46.2 GB
- Zero new parameters (pure scheduling)
- Temperature released at epoch 80/60 for free learning thereafter

### What happened

The temperature curriculum consistently worsened p_tan and p_oodc relative to baseline. The broad→sharp routing hypothesis did not help the model generalize better to tandem/OOD configurations.

**Why it may have failed:**

1. **High initial temperature is harmful, not helpful.** Starting at temp=2.0 (very broad, near-uniform softmax) forces the model to spread signal across all slices uniformly for the first 80 epochs. This prevents the slices from specializing early — and tandem transfer requires correctly routing geometry features (gap/stagger via GSB) to the right slices from the start.

2. **Interferes with GSB and existing routing.** The GSB () already helps the model route tandem vs. non-tandem samples to different slices. Forcing high temperature early disrupts this mechanism, potentially causing the model to unlearn good routing patterns it would otherwise develop naturally.

3. **Config B is marginally better than Config A** (29.35 vs 29.75 p_tan), consistent with the shorter/milder annealing being less disruptive, but neither beats baseline.

4. **p_oodc regression** (~+5%) suggests that the curriculum disrupts the model's ability to handle extreme conditions, not just tandem transfer.

### Suggested follow-ups

1. **Temperature scheduling is unlikely to help in this direction.** The result is clear: externally controlling attention temperature hurts. The model finds better routing on its own.

2. **Lower initial temp might be interesting:** Instead of starting high and going low, testing  in reverse (start at baseline ~0.5, anneal to 0.25 to force specialization) or just keeping temp fixed lower than default could be explored.

3. **Focus on other mechanisms** — data augmentation targeting tandem geometry diversity, or architectural improvements to cross-foil attention, seem more promising.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2174 because ADVISOR: Closing — both temperature curriculum configs regress p_tan and p_oodc.

**Config A** (2.0→0.3, 80ep): p_tan=29.79 (+4.2%), p_oodc=8.14 (+5.7%)
**Config B** (1.5→0.5, 60ep): p_tan=29.38 (+2.7%), p_oodc=8.12 (+5.5%)

Your analysis is spot on: high initial temperature disrupts early slice specialization, and the GSB routing mechanism needs early accurate routing to function properly. Forcing near-uniform attention in the first 80 epochs wastes critical learning time.

**Key insight:** The Transolver's attention routing is already near-optimal without external scheduling. Any mechanism that forces broad attention early just delays specialization. This joins the pattern of convergence dynamics being absorbed by the EMA + cosine schedule.

Thank you for the thorough analysis and clean follow-up suggestions..
